### PR TITLE
GC: Diffs on io

### PR DESCRIPTION
**Pull request by Ngoguey42**
This PR targets the GC branch.

This addresses @tomjridge comments left in the code

### CODE REVIEW

File: src/irmin-pack/unix/io.ml [hunk 1]
```diff
@@ -108,154 +116,114 @@ module Unix = struct
...
+                 has a handle on it, so it cannot be mutated; so
+                 [Bytes.unsafe_to_string] is obviously safe here *)
+              Ok (Bytes.unsafe_to_string buf)
+        with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2)))
 
-  (* NOTE copied from read_to_string above; ensure the code for both functions remains in
-     sync *)
   let read_exn t ~off ~len buf : unit =
-    assert (len <= Bytes.length buf);
-    try
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/io.ml:221]: The API I designed was not right. It did not reflect the fact that we should not open `try` clauses in the `_exn` functions, for performance reasons. The `Unix.Unix_error` exceptions should be propagated up to the irmin user.

File: src/irmin-pack/unix/io.ml [hunk 2]
```diff
@@ -63,43 +62,52 @@ module Unix = struct
...
-  let default_create_perm = 0o660
-  let default_open_perm = 0o660
-  let default_mkdir_perm = 0o770 (* or 0o777? *)
+  let default_create_perm = 0o644
+  let default_open_perm = 0o644
+  let default_mkdir_perm = 0o755
 
-  (* FIXME meaning of overwrite? we assume it means "overwrite: if the file exists,
-     truncate it and use the truncated file instead"; otherwise, if overwrite=false, and
-     we try to create a file that already exists, we fail with EEXIST *)
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/io.ml:94]: Truncating seems OK

File: src/irmin-pack/unix/io.ml [hunk 3]
```diff
@@ -63,43 +62,52 @@ module Unix = struct
...
+    | `No_such_file_or_directory
+    | `Invalid_parent_directory ]
 
-  type move_file_error = [ `Io_misc of misc_error ]
+  include Errors
 
-  (* the interface indicates that we need to catch double close errors; so type t needs
-     some field to record whether the instance has been closed; we also need a field for
-     "readonly" (to avoid querying the fd for this info); also for path (because of "path"
-     function below) *)
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/io.ml:75]: 👍

**Review by icristescu — APPROVED** [review]
LGTM!

**Pull request merged**